### PR TITLE
Entity manger adaptation

### PR DESCRIPTION
**Pull request by nest1ing**
These changes:
 - bring a support of any inventory managers including `phosphor-inventory-manager`, `entity-manager` and so on.
 - replace the `json-c` usage with a `nlohmann/json` as it's more preferred by the upstream.
 - add support for D-Bus on a remote host, useful during debug.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -14,21 +14,23 @@
...
-    printf("Copyright (c) 2020 YADRO.\n");
+    printf("Copyright (c) 2021 YADRO.\n");
     printf("Version " VERSION "\n");
     printf("Usage: %s [OPTION...]\n", app);
     printf("  -n, --name=NAME  Print item with specified name only\n");
     printf("  -a, --all        Also print non-present units\n");
     printf("  -e, --empty      Also print empty properties\n");
     printf("  -j, --json       Print in JSON format\n");
     printf("  -h, --help       Print this help and exit\n");
+    printf("  -H, --host       Get data from remote host over SSH\n");
```
Comment (artemsen) [inline src/main.cpp:25]: Why do we need this? To read inventory from another BMC?

File: src/printer.cpp [hunk 2]
```diff
@@ -77,48 +77,49 @@ void Printer::printJson(const std::vector<InventoryItem>& items) const
...
+            nlohmann::json jsonProp;
 
             // get value from variant
             std::visit(
                 [&jsonProp](auto&& arg) {
                     using T = std::decay_t<decltype(arg)>;
                     if constexpr (std::is_same_v<T, bool>)
-                        jsonProp = json_object_new_boolean(arg);
+                    {
+                        jsonProp = arg;
```
Comment (artemsen) [inline src/printer.cpp:93]: Now we don't need to have "if constexpr".

File: src/printer.cpp [hunk 3]
```diff
@@ -77,48 +77,49 @@ void Printer::printJson(const std::vector<InventoryItem>& items) const
...
-        json_object_object_add(json, item.name.c_str(), jsonItem);
+        json.emplace(item.name, jsonItem);
     }
 
-    printf("%s\n",
-           json_object_to_json_string_ext(json, JSON_C_TO_STRING_PRETTY |
-                                                    JSON_C_TO_STRING_SPACED));
-
-    json_object_put(json);
+    printf("%s\n", json.dump(2).c_str());
```
Comment (artemsen) [inline src/printer.cpp:122]: What is "2"?

File: src/main.cpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2020 YADRO
+// Copyright (C) 2021 YADRO
```
Comment (artemsen) [inline src/main.cpp:2]: It is a bad practice to change the date in copyrights.

https://stackoverflow.com/questions/2390230/do-copyright-dates-need-to-be-updated:
The copyright notice on a work establishes a claim to copyright. The date on the notice establishes how far back the claim is made. This means if you update the date, you are no longer claiming the copyright for the original date and that means if somebody has copied the work in the meantime and they claim its theirs on the ground that their publishing the copy was before your claim, then it will be difficult to establish who is the originator of the work.

File: meson_options.txt [hunk 5]
```diff
@@ -13,17 +13,13 @@ option('mapper-iface',
        description: 'Object mapper interface')
 
 # Inventory manager
-option('inventory-service',
```
Comment (artemsen) [inline meson_options.txt:16]: What about Nicole/Vesnin?

File: src/main.cpp (hunk 1 above)
Comment (ikmsk10) [reply src/main.cpp:25]: I guess that is to have a way to debug `lsinventory` on the developer station and retrieve data from a real BMC stand.
But I've agreed with Artem, it would be better to define a compile-time flag that is able to disable that option in the production.

Comment (nest1ing) [reply src/main.cpp:27]: moved under `#ifdef` and configuration option

File: src/printer.cpp (hunk 2 above)
Comment (nest1ing) [reply src/printer.cpp]: An empty string is not an empty JSON object. ;)
I've moved the empty string checks outside of visitor.

File: src/printer.cpp (hunk 3 above)
Comment (nest1ing) [reply src/printer.cpp]: fixed

File: src/main.cpp (hunk 4 above)
Comment (nest1ing) [reply src/main.cpp]: reverted

File: meson_options.txt (hunk 5 above)
Comment (nest1ing) [reply meson_options.txt:16]: PIM also creates the objects with required interfaces. So, it works properly on those platforms too.

**Review by ikmsk10 — DISMISSED** [review]

File: test/inventory_test.cpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2020 YADRO
+// Copyright (C) 2021 YADRO
```
Comment (alatarum) [inline test/inventory_test.cpp:2]: 2020-2021?

File: meson_options.txt [hunk 7]
```diff
@@ -13,17 +13,13 @@ option('mapper-iface',
...
-       type: 'string',
-       value: 'xyz.openbmc_project.Inventory.Manager',
-       description: 'Inventory service')
 option('inventory-path',
        type: 'string',
        value: '/xyz/openbmc_project/inventory',
        description: 'Root path of inventory objects')
 option('inventory-iface',
        type: 'string',
-       value: 'xyz.openbmc_project.Inventory.Item',
```
Comment (alatarum) [inline meson_options.txt:26]: why did you change this?
IMHO would be better to fix inventory daemons to provide ".Inventory.Item" for every item.

File: test/inventory_test.cpp (hunk 6 above)
Comment (nest1ing) [reply test/inventory_test.cpp:2]: Ups... Forgot to revert this.

File: meson_options.txt (hunk 7 above)
Comment (nest1ing) [reply meson_options.txt:26]: `xyz.openbmc_project.Inventory.Decorator.Asset` provides the most usable fields like `Manufacturer`, `PartNumber`, `SerialNumber` and it's implemented in all objects that should be shown.
`xyz.openbmc_project.Inventory.Item` provides `PrettyName` and `Presence` and not every object implements them.
Likely this issue should be fixed in corresponding services, but `Decorator.Asset` seems to me more reliable.

File: test/inventory_test.cpp (hunk 6 above)
Comment (nest1ing) [reply test/inventory_test.cpp]: reverted

**Review by artemsen — APPROVED** [review]

**Review by alatarum — APPROVED** [review]

**Review by AlexanderAmelkin — APPROVED** [review]
